[Core: Login, User Accounts] Allow certain users to access the system in a time window basis

racostas opened the pull request:
### Brief summary of changes

For the project we are working at CHU Sainte-Justine we realize we will need to set (for a given user) access to the system but only for a given time window (due to policies of data access and privacy). It's a specific requirement of our project that could be useful for others. 
The way it was implemented (following @ridz1208 recommendations. Very helpful!! Thanks!) should not break/interfere others functionalities of the system. 

The changes are all additions limited to the following files:

php/libraries/SinglePointLogin.class.inc
modules/user_accounts/templates/form_edit_user.tpl
modules/user_accounts/php/edit_user.class.inc

Also it's necessary to add two fields on the Users table:

SQL/New_patches/2018-08-22_user_activeFrom-activeTo.sql

### This resolves issue...

- [ ] Redmine? # -

- [ ] Github? # -

### To test this change...

1. To access the system with a user with permissions to add/modify another.
2. To modify the user access time windows filling the new fields 'Active from' and/or 'Active to' that should shows up between the fields 'Active' and 'Pending approval'. NULLs are allowed, which will mean that there is not restrictions.
3. According to the dates set the user modified should not be able to access the system (due to expired/no yet active user).

![image](https://user-images.githubusercontent.com/37309344/44554134-c6614400-a6fd-11e8-92d6-80934e8b2af3.png)

![image2](https://user-images.githubusercontent.com/37309344/44554143-ceb97f00-a6fd-11e8-88af-9c946a970dba.png)

All suggestions/modifications are very welcome! Thanks.
CSJ team.

driusan commented:
@racostas This is failing Travis because it needs to have phpcs run on it before Travis runs the test suite. The test suite runs:

```
vendor/bin/phpcs --standard=docs/LorisCS.xml php/libraries php/exceptions php/installer;
vendor/bin/phpcs --standard=docs/LorisCS.xml --extensions=php/php htdocs;
vendor/bin/phpcs --standard=docs/LorisCS.xml --extensions=php/php,inc/php modules/;
```

from the root of the LORIS directory as a first step. (You can replace "phpcs" with "phpcbf" to fix the ones reported that have an `[x]` beside them, but the remaining ones like line length too line need to be fixed manually. You can see the log from Travis CI by clicking on the red x beside a commit id in GitHub)

driusan commented:
I don't have any feelings one way or another over to/until and I doubt it's possible to make it work without the alter table since the query needs to select the columns.

johnsaigle commented on modules/user_accounts/php/edit_user.class.inc:
```diff
@@ -1086,6 +1113,18 @@ class Edit_User extends \NDB_Form
+        //======================================
+
+	if(($values['active_to'] != NULL) && ($values['active_from'] > $values['active_to'])){
+		$errors['active_timeWindows'] = "Please notice that the <i>\"Active from\"</i> date should be lesser or equal to the <i>\"Active to\"</i> date.";
```
We usually don't inject HTML directly into error messages like this as a matter of policy. If the italics are very important to you here, this should be done in the Smarty template instead.

cmadjar commented:
@johnsaigle I like your suggestion!

johnsaigle commented on php/libraries/SinglePointLogin.class.inc:
```diff
@@ -328,6 +331,26 @@ class SinglePointLogin
 
+		// check if the account is longer active
+		$date = new DateTime();
+		$currentDay = $date->getTimestamp();
```
Technically this should be named something like `$currentTime` as it will incorporate the hour, minute, second, etc.

johnsaigle commented on php/libraries/SinglePointLogin.class.inc:
```diff
@@ -328,6 +331,26 @@ class SinglePointLogin
+
+		// check if the account is active		
+		if(($row['active_from'] != NULL) && ($currentDay < strtotime($row['active_from']))){
+			$this->_lastError = "Your account is not active yet. According to our records it will be active from ".$row['active_from']."."
```
You're already using double-quotes here so I think string interpolation is a bit cleaner than concatenating.

```php
$this->_lastError = "Your account is not active yet. According to our records it will be active from {$row['active_from']}."
```

racostas commented:
Hi @driusan , Thanks for the comments. They are very appreciated!

I did the phpcs and phpbf for the files I modified (php/libraries/SinglePointLogin.class.inc,  modules/user_accounts/templates/form_edit_user.tpl, modules/user_accounts/php/edit_user.class.inc). After that and some manual fix that was required all seems to be OK when I run phpcs again for this three files. What is strange now is that Travis is giving me format errors(specifically warnings) in files I have not touched at all? Maybe I’m missing something obvious??

The files I modified do not seem to be among the ones reported with issues now. But maybe because they did not got analyzed as it seems that Travis exceed the execution time. 

The final result is that Travis gives the test as failed.

![screenshot_1](https://user-images.githubusercontent.com/37309344/45003330-58dfce00-afaf-11e8-8f0b-4adeec76b2c7.png)

![screenshot_2](https://user-images.githubusercontent.com/37309344/45003336-62693600-afaf-11e8-8606-8f781a02ef4d.png)

Thanks in advance for the help you may provide us,
CNBP Team.

driusan commented:
@racostas The warnings are from the equivalent of phpcs for javascript and are harmless, they don't fail the build. Those warnings are printed on every pull request. The "E" lines are from the test suite and what matters, but since it's timing out before finishing you're not getting the summary of what they mean that's usually printed after it finishes, which makes it hard to say exactly what's wrong...

Since every test is failing and you're changing the login, I'd guess that something is broken with the login process under Travis causing every test to fail when it tries logging in. It could be because you didn't update the schema file, or because the test data needs to be updated to insert properly with the schema changes. @kongtiaowang might be able to help you get to the bottom of it.

kongtiaowang commented:
https://github.com/aces/Loris/pull/3916

racostas commented:
Hi @kongtiaowang. Travis test passed now thank to your insights. Great help! Thank you very much!

driusan commented:
@cmadjar Is your review still valid or has it been adequately addressed?
